pkg/endpointmanager/manager.go

```diff
@@ -627,7 +627,17 @@ func (mgr *endpointManager) expose(ep *endpoint.Endpoint) error {
 // manager.
 func (mgr *endpointManager) RestoreEndpoint(ep *endpoint.Endpoint) error {
 	ep.SetDefaultConfiguration(true)
-	return mgr.expose(ep)
+	err := mgr.expose(ep)
+	if err != nil {
+		return err
+	}
+	mgr.mutex.RLock()
```

Comment:
Probably unlikely, but because all endpoint subscriber events are done in a separate critical section from expose/unexpose, could there be any issues if say a restore/delete event where emitted out of order?

(I think this is fine in your case since the other hooks appear to just be stubs)

Reply:
Seems dangerous to hold the RLock() and calling `EndpointRestored` here. At the very least it looks redirect policy manager is calling `localPods.Store()` which blocks until we've pulled the pods from api-server and that may delay endpoint restoration. 

It would be better to grab a copy of `subscribers` with RLock, then release it and call the subscribers. This does of course mean that a subscriber may get called after it has unsubscribed, but that seems like a lesser evil.

Reply:
Maybe... AFAICS, `unsubscribe` is called in the tear down phase, so it may not race with the restore path.

Reply:
Revised this logic in the latest revision.